envoy/network/address.h

```diff
@@ -230,6 +231,32 @@ class Instance {
    * @return SocketInterface to be used with the address.
    */
   virtual const Network::SocketInterface& socketInterface() const PURE;
+
+  /**
+   * For support the absl::Hash.
+   */
+  virtual void abslHashValue(absl::HashState state) const PURE;
```

Comment:
> My concern with this though is this may cause different hashing of upstream connections than we currently have and thrashing during rollout if folks expect consistent hashing across instances. cc @htuch @mattklein123 for if this is an actual problem and if so how we would do it safely.

The fact that we are already using absl::Hash for this means that there is no guarantee of stability between restarts (though in practice stability is likely), so it's probably OK to make this change. If we really want stability in this path we should be using one of the stable hashes in here: https://github.com/envoyproxy/envoy/blob/main/source/common/common/hash.h. (Like we do for the consistent hash load balancers for this reason.)

So I guess this is probably OK? Perhaps we should open an issue to migrate this use of hash to a stable hash though?

Reply:
I'd think absl::Hash should be pretty stable across restarts for the same IP address and backend sets. If we think stability is currently likely and this may change it, how about a runtime guard and release note as a compromise?  We could also then leave the guard false until all the code is guarded, so have reviewable code segments but not have inconsistent hashing within the binary.

Reply:
Thanks for raising this @alyssawilk. 

I think the PR is mainly affecting how we hash for STL/ABSL data structures, rather than in places like `hash_policy.cc`, where presumably we still keep using string representation. I think this should also hold for the affinity balancers and ring construction. Maybe I'm missing some place where this matters though, so it can't hurt to runtime guard.

Reply:
Actually I hash on the InstanceConstSharePtr, that why I added `struct AddressKey` below https://github.com/envoyproxy/envoy/pull/24198/files/97988a05a5450639759221f626aac84a7b9d2a64#diff-131ae3cdf1b8289789a6da638553a5748c9c007d28ff04a3f3c8fcb4f533e58cR245

So I guess any code without explicitly using the `AddressKey`, then won't use this hash code, that means all the existing code won't be changed by this PR?

Reply:
absl::Hash is not deterministic across process invocations. It uses ASLR as a seed to randomize hash generation. https://github.com/abseil/abseil-cpp/blob/570dc02b0b70bd4fb37efd79021890fa9edc283f/absl/hash/internal/hash.h#L1120

So changing the input into the hash function should not matter with respect to hash stability across process restarts.